File: leetcode/fast-and-slow-pointer/0202_happy_number.cpp
// LC 202. Happy Number | Easy
class Solution {
    int nxt(int n) { int s = 0; while (n) { int d = n % 10; s += d*d; n /= 10; } return s; }
public:
    bool isHappy(int n) {
        int slow = n, fast = nxt(n);
        while (fast != 1 && slow != fast) { slow = nxt(slow); fast = nxt(nxt(fast)); }
        return fast == 1;
    }
};

File: leetcode/fast-and-slow-pointer/0202_happy_number.py
# LC 202. Happy Number | Easy
class Solution:
    # Floyd's cycle detection
    def isHappy(self, n):
        def nxt(x):
            s = 0
            while x: x, d = divmod(x, 10); s += d * d
            return s
        slow, fast = n, nxt(n)
        while fast != 1 and slow != fast:
            slow = nxt(slow); fast = nxt(nxt(fast))
        return fast == 1
